taichi/program/program_impl.h

```diff
@@ -55,6 +56,14 @@ class ProgramImpl {
    */
   virtual std::unique_ptr<AotModuleBuilder> make_aot_module_builder() = 0;
 
+  virtual Device *get_compute_device() {
+    return nullptr;
+  }
+
```

Comment:
Just for my own naive understanding ;), these two functions are not pure virtual because they don't have to be implemented or?

Reply:
I didn't make them pure virtual because for now returning a `nullptr` is a reasonable default behaviour. Currently, not every backend has a `Device` in their `ProgramImpl`.